### PR TITLE
Remove unnecessary modulo operations from index_to_linear of cluster_mesh

### PR DESCRIPTION
**Pull request by Wentzell**
* Explicitly check is_within_boundary for index_to_linear and index_to_point in Debug mode
* Print stack_trace in contract macros EXPECTS, ENSURES, ASSERT and disable without debugging
* Minor formatting corrections

### CODE REVIEW

File: test/triqs/gfs/gf_empty.cpp [hunk 1]
```diff
@@ -9,6 +9,6 @@ TEST(Gf, Exceptions) {
 
   auto g = gf<imtime, scalar_valued>{};
 
-  EXPECT_THROW(g[closest_mesh_pt(5.3)], triqs::arrays::key_error);
+  ASSERT_DEATH(g[closest_mesh_pt(5.3)], "");
```
Comment (parcollet) [inline test/triqs/gfs/gf_empty.cpp:12]: why this change ?

File: triqs/gfs/meshes/discrete.hpp [hunk 2]
```diff
@@ -51,9 +51,19 @@ namespace triqs {
       ///
       utility::mini_vector<size_t, 1> size_of_components() const { return {size()}; }
 
+      /// Is the point in the mesh ?
+      bool is_within_boundary(all_t) const { return true; }
```
Comment (parcollet) [inline triqs/gfs/meshes/discrete.hpp]: by definition all_t is within mesh !
We should not have to implement this in **each** mesh, it should be intercepted
at the caller

File: triqs/gfs/meshes/imfreq.hpp [hunk 3]
```diff
@@ -134,14 +134,26 @@ namespace triqs::gfs {
     ///
     utility::mini_vector<size_t, 1> size_of_components() const { return {size_t(size())}; }
 
+    /// Is the point in mesh ?
+    bool is_within_boundary(all_t) const { return true; }
```
Comment (parcollet) [inline triqs/gfs/meshes/imfreq.hpp]: same

File: triqs/gfs/meshes/imtime.hpp [hunk 4]
```diff
@@ -55,7 +55,9 @@ namespace triqs {
       gf_mesh(double beta, statistic_enum S, long n_time_slices) : gf_mesh({beta, S}, n_time_slices) {}
 
       /// For imtime the point is always in the mesh, since we use anti-periodicity or periodicity. Needed for cartesian product.
-      bool is_within_boundary(double x) const { return true; }
+      static constexpr bool is_within_boundary(all_t) { return true; }
```
Comment (parcollet) [inline triqs/gfs/meshes/imtime.hpp]: Really ? I thought we had removed the periodicity in evaluation ??
To be discussed.

File: triqs/gfs/meshes/linear.hpp [hunk 5]
```diff
@@ -55,22 +66,22 @@ namespace triqs {
 
       utility::mini_vector<size_t, 1> size_of_components() const { return {size_t(size())}; }
 
+      /// Is the point in mesh ?
+      bool is_within_boundary(all_t) const { return true; }
```
Comment (parcollet) [inline triqs/gfs/meshes/linear.hpp]: same

File: triqs/lattice/cluster_mesh.hpp [hunk 6]
```diff
@@ -118,17 +118,21 @@ namespace triqs {
       using index_t        = utility::mini_vector<long, 3>;
       using linear_index_t = long;
 
+      /// Reduce index modulo to the lattice.
```
Comment (parcollet) [inline triqs/lattice/cluster_mesh.hpp:121]: What about the 
modulo_reduce
at the end of the file and used in the mesh_point ?

**Review by parcollet — APPROVED** [review]
two main points to discuss : 
- g(tau) periodicity
- modulo_reduce

File: test/triqs/gfs/gf_empty.cpp (hunk 1 above)
Comment (Wentzell) [reply test/triqs/gfs/gf_empty.cpp:12]: This change is necessary because `g[closest_mesh_pt(5.3)]` will now immediately violate the precondition `EXPECTS(_mesh.is_within_boundary(arg));`. I.e. the test can no longer check for an exception but has to assert that the program terminates.

File: triqs/gfs/meshes/discrete.hpp (hunk 2 above)
Comment (Wentzell) [reply triqs/gfs/meshes/discrete.hpp]: As agreed we delay this interception until after the 2.2 release.

File: triqs/lattice/cluster_mesh.hpp (hunk 6 above)
Comment (Wentzell) [reply triqs/lattice/cluster_mesh.hpp:121]: I have removed the `modulo_reduce` as it achieves the same goal as `index_modulo`

File: triqs/gfs/meshes/imtime.hpp (hunk 4 above)
Comment (Wentzell) [reply triqs/gfs/meshes/imtime.hpp]: This is in fact a bug! I have pushed a fix. We are now falling back to the `is_within_boundary` implementations of the `linear_mesh`

File: triqs/lattice/cluster_mesh.hpp [hunk 7]
```diff
@@ -257,21 +254,17 @@ namespace triqs {
...
       double operator[](int d) const { return operator()(d); }
       friend std::ostream &operator<<(std::ostream &out, mesh_point const &x) { return out << (lattice_point)x; }
-      mesh_point operator-() const { return mesh_point{*m, m->modulo_reduce({-index()[0], -index()[1], -index()[2]})}; }
+      mesh_point operator-() const { return mesh_point{*m, m->index_modulo({-index()[0], -index()[1], -index()[2]})}; }
       mesh_t const &mesh() const { return *m; }
     };
 
     // --- impl
-    inline mesh_point<cluster_mesh> cluster_mesh::operator[](index_t i) const { return mesh_point<cluster_mesh>{*this, this->modulo_reduce(i)}; }
+    inline mesh_point<cluster_mesh> cluster_mesh::operator[](index_t i) const { return mesh_point<cluster_mesh>{*this, this->index_modulo(i)}; }
```
Comment (parcollet) [inline triqs/lattice/cluster_mesh.hpp]: why do we keep the modulo here ??

Comment (Wentzell) [reply triqs/lattice/cluster_mesh.hpp]: That is a good question. I just did the renaming in the file.
We should just use `i` here protect it with `EXPECTS(is_within_boundary(i))`. Do you agree?
